### ISSUE TITLE
Request for Support of Mixed Precision Training

### ISSUE THREAD

**Issue by JAMESKirkQI** [TorchJD/torchjd#244]
**Description:**  
Hi, thank you for your great work on TorchJD!  

I would like to request support for mixed precision training in TorchJD. Currently, using `torch.amp.GradScaler` for loss scaling is a common approach to enable efficient half-precision training on GPUs. However, it seems that TorchJD does not natively support this yet.  

Would it be possible to integrate mixed precision training into TorchJD? For example, adapting the optimizer step to work with `torch.amp.GradScaler` as follows:  

```python
from torch.amp import GradScaler  
    shared_module = Sequential(Linear(10, 5), ReLU(), Linear(5, 3), ReLU()).cuda()
    task1_module = Linear(3, 1).cuda()
    task2_module = Linear(3, 1).cuda()
    params = [
        *shared_module.parameters(),
        *task1_module.parameters(),
        *task2_module.parameters(),
    ]
    scaler = GradScaler('cuda')
    loss_fn = MSELoss()
    optimizer = SGD(params, lr=0.1)
    aggregator = UPGrad()

    inputs = torch.randn(8, 16, 10).cuda()  # 8 batches of 16 random input vectors of length 10
    task1_targets = torch.randn(8, 16, 1).cuda()  # 8 batches of 16 targets for the first task
    task2_targets = torch.randn(8, 16, 1).cuda()  # 8 batches of 16 targets for the second task

    for input, target1, target2 in zip(inputs, task1_targets, task2_targets):
        with torch.autocast(device_type='cuda', dtype=torch.float16):
            features = shared_module(input)
            output1 = task1_module(features)
            output2 = task2_module(features)
            loss1 = loss_fn(output1, target1)
            loss2 = loss_fn(output2, target2)

            optimizer.zero_grad()
            mtl_backward(losses=[loss1, loss2], features=features, aggregator=aggregator)
            scaler.step(optimizer)
```

This feature would be beneficial for training efficiency and memory savings. Please let me know if there's anything I can do to help.  

Thanks!

**Comment by ValerianRey** [comment]
Hi! Thanks for your interest and for helping improve torchjd!

I think that there are some problems with your code example. First, the documentation of [torch.amp.autocast](https://pytorch.org/docs/stable/amp.html#torch.autocast) says:
> [autocast](https://pytorch.org/docs/stable/amp.html#torch.autocast) should wrap only the forward pass(es) of your network, including the loss computation(s). Backward passes under autocast are not recommended. Backward ops run in the same type that autocast used for corresponding forward ops.

So the lines:
```python
optimizer.zero_grad()
mtl_backward(losses=[loss1, loss2], features=features, aggregator=aggregator)
scaler.step(optimizer)
```
should be moved out of the `torch.autocast` context.

Second, you should call the `scale` method of your scaler to scale the losses, and call its `update` method at each iteration to update the scaling factor. See [this](https://github.com/pytorch/pytorch/blob/v2.6.0/torch/amp/grad_scaler.py#L49-L117) for the precise documentation of the `GradScaler` class, and [this](https://pytorch.org/docs/stable/notes/amp_examples.html#typical-mixed-precision-training) for a usual example.

When using torchjd, the only thing that changes is that instead of calling torch's backward method with `scaler.scale(loss).backward()`, you should call torchjd's `backward` or `mtl_backward` method on the scaled losses, with:
```python
scaled_loss1, scaled_loss2 = scaler.scale([loss1, loss2])
mtl_backward(losses=[scaled_loss1, scaled_loss2], features=features, aggregator=aggregator)
```


With these fixes (+ the full imports), your example becomes:

```python
import torch
from torch.amp import GradScaler
from torch.nn import Sequential, Linear, ReLU, MSELoss
from torch.optim import SGD

from torchjd import mtl_backward
from torchjd.aggregation import UPGrad

shared_module = Sequential(Linear(10, 5), ReLU(), Linear(5, 3), ReLU()).cuda()
task1_module = Linear(3, 1).cuda()
task2_module = Linear(3, 1).cuda()
params = [
    *shared_module.parameters(),
    *task1_module.parameters(),
    *task2_module.parameters(),
]
scaler = GradScaler('cuda')
loss_fn = MSELoss()
optimizer = SGD(params, lr=0.1)
aggregator = UPGrad()

inputs = torch.randn(8, 16, 10).cuda()  # 8 batches of 16 random input vectors of length 10
task1_targets = torch.randn(8, 16, 1).cuda()  # 8 batches of 16 targets for the first task
task2_targets = torch.randn(8, 16, 1).cuda()  # 8 batches of 16 targets for the second task

for input, target1, target2 in zip(inputs, task1_targets, task2_targets):
    with torch.autocast(device_type='cuda', dtype=torch.float16):
        features = shared_module(input)
        output1 = task1_module(features)
        output2 = task2_module(features)
        loss1 = loss_fn(output1, target1)
        loss2 = loss_fn(output2, target2)
    
    # De-indented
    scaled_loss1, scaled_loss2 = scaler.scale([loss1, loss2])  # Added
    optimizer.zero_grad()
    mtl_backward(losses=[scaled_loss1, scaled_loss2], features=features, aggregator=aggregator)  # Changed
    scaler.step(optimizer)
    scaler.update()  # Added
```

I tested this exact code and it seems to work. I didn't test this on the full training of a neural network though.

I think that the Jacobian computed during the call to `mtl_backward` will always be of `float32` type (or whatever type your model was on), so the aggregation will work the same as if no AMP was used, and the `.grad` fields will be filled with `float32` tensors, exactly like `torch` itself behaves (with and without AMP). So normally this should do exactly what you want.

@JAMESKirkQI Could you try out the changes that I suggested and update us about whether it works for your original use-case (I'm assuming you're trying to train real neural networks with this)?

@PierreQuinton This is the second time someone asks about AMP (see #135), so I suggest we add this as a usage example in the documentation. What do you think?

**Comment by JAMESKirkQI** [comment]
Thank you for your prompt response! The example code you provided is very helpful, and after adapting it, it perfectly covers my original use case. I really appreciate your support and effort in improving TorchJD.

Thanks again!
